src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -697,28 +697,43 @@ putPaddingTo64 written
 fromSBS :: ShortByteString -> Either String (NumEntries, CompactIndex)
 fromSBS (SBS ba') = do
     let ba = ByteArray ba'
-    let len = sizeofByteArray ba
-    when (mod8 len /= 0) $ Left "Length is not multiple of 64 bit"
-    when (len < 36) $ Left "Too small for header and footer"
+    let len8 = sizeofByteArray ba
+    when (mod8 len8 /= 0) $ Left "Length is not multiple of 64 bit"
+    when (len8 < 36) $ Left "Input is too short for header and footer"
 
+    -- check version
     let version = indexByteArray ba 0 :: Word32
     when (version == byteSwap32 indexVersion) $ Left "Non-matching endianness"
     when (version /= indexVersion) $ Left "Unsupported version"
 
-    let len64 = div8 len
-    let rfprec = fromIntegral (indexByteArray ba (len64 - 3) :: Word64)
-    let numPages = fromIntegral (indexByteArray ba (len64 - 2) :: Word64)
-    let numEntries = fromIntegral (indexByteArray ba (len64 - 1) :: Word64)
+    -- read footer
+    let len64 = div8 len8
+    let getPositive offset64 = do
+          let w = indexByteArray ba offset64 :: Word64
+          when (w > fromIntegral (maxBound :: Int)) $
+            Left "Size information is too large for Int"
+          return (fromIntegral w)
 
+    rfprec <- getPositive (len64 - 3)
+    numPages <- getPositive (len64 - 2)
+    numEntries <- getPositive (len64 - 1)
```

Comment:
Similar idea to my previous comment: if there is a large slice of the bytearray that is unused between the data and the footer, then we're keeping those bytes live. Maybe we need an exact check of file length vs. these parameters?

Reply:
(done)